content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -1,38 +1,408 @@
+
+.. code-block:: ada
+
+  Units : constant Integer := (X'Size + Storage_Unit - 1) / Storage_Unit;
+
+Remember that `'Size` returns a value in terms of bits. There are more
+direct ways to determine that size information but this will serve as an
+example of the sort of thing you might do with that constant.
+
+`Word_Size` is the number of bits in the units that the machine natively
+manipulates. On a 32-bit machine we'd expect Word to have a value of 32;
+on a 64-bit machine it would probably be 64, and so on.
+
+`Storage_Unit` and `Word_Size` are obviously related. For this Arm target a
+word consists of four eight-bit storage elements.
+
+Another frequently referenced declaration in package System is that of
+the type representing memory addresses, along with a constant for the
+null address designating no storage element.
+
+.. code-block:: ada
+
+   type Address is private;
+   Null_Address : constant Address;
+
+You may be wondering why type Address is a private type, since that
+choice means that we programmers cannot treat it like an ordinary
+(unsigned) integer value. Portability is of course the issue, because
+addressing, and thus address representation, varies among computer
+architectures. Not all architectures have a flat address space directly
+referenced by numeric values, although that is common. Some are
+represented by a base address plus an offset, for example. Therefore,
+the representation for type Address is hidden from us, the clients.
+Consequently we cannot simply treat address values as numeric values.
+Don't worry, though. The operations we need are provided.
+
+Package `System` declares these comparison functions, for example:
+
+.. code-block:: ada
+
+   function "<"  (Left, Right : Address) return Boolean;
+   function "<=" (Left, Right : Address) return Boolean;
+   function ">"  (Left, Right : Address) return Boolean;
+   function ">=" (Left, Right : Address) return Boolean;
+   function "="  (Left, Right : Address) return Boolean;
+
+These functions are intrinsic, i.e., built-in, meaning that the compiler
+generates the code for them directly at the point of calls. There is no
+actual function body for any of them so there is no performance penalty.
+
+Any private type directly supports the equality function, and
+consequently the inequality function, as well as assignment. What we
+don't get here is address arithmetic, again because we don't have a
+compile-time view of the actual representation. That functionality is
+provided by package `System.Storage_Elements`, a child package we will
+cover later. We should say though, that the need for address arithmetic
+in Ada is rare, especially compared to C.
+
+Having type `Address` presented as a private type is not, strictly
+speaking, required by the language. Doing so is a good idea for the
+reasons given above, and is common among vendors. Not all vendors do,
+though.
+
+Note that `Address` is the type of the result of the query attribute
+`Address`.
```

Comment:
```suggestion
Note that :ada:`Address` is the type of the result of the query attribute
:ada:`Address`.
```

Reply:
Also, it would be good to have a short code example that illustrates this statement. For example:

```
.. code:: ada compile_button project=Courses.Intro_To_Embedded_Sys_Prog.Low_Level_Programming.Address_Attribute

    with System; use System;

    procedure Main is
       X : Integer;
       A : Address := X'Address;
    begin
       null;
    end Main;
```